baseplate/context/cassandra.py

```diff
@@ -150,6 +152,71 @@ def execute_async(self, query, parameters=None, timeout=_NOT_SET):
 
     def prepare(self, query):
         trace_name = "{}.{}".format(self.context_name, "prepare")
-        with self.server_span.make_child(trace_name) as span:
-            span.set_tag("statement", query)
+        with self._get_span(trace_name, query):
             return self.session.prepare(query)
+
+    def _get_span(self, trace_name, query):
+        """Get trace span for query.
+
+        Metadata from the passed query is extracted and the created span is
+        tagged accordingly.
+
+        """
+        span = self.server_span.make_child(trace_name)
+        cql_str = None
+        if isinstance(query, string_types):
+            cql_str = query
+        elif isinstance(query, (SimpleStatement, PreparedStatement)):
+            cql_str = query.query_string
+        elif isinstance(query, BoundStatement):
+            cql_str = query.prepared_statement.query_string
+
+        # TODO: include custom payload
+        span.set_tag("statement", cql_str)
+        for key, value in CQLMetadataExtractor.extract(cql_str).items():
+            span.set_tag(key, value)
+
+        return span
+
+
+class CQLMetadataExtractor(object):
```

Comment:
:grimacing: I'm really really skeptical of regex parsing arbitrary CQL queries for this. Do we have any data on the performance impact of doing this?

Does extracting from the CQL like this actually help when we have the full statement in the tags anyway?

Reply:
> Does extracting from the CQL like this actually help when we have the full statement in the tags anyway?

I'm not sure of the desire behind the ticket. My guess is wanting the spans from baseplate to resemble those generated by r2. If people are searching Zipkin traces by column family tag names, this would be helpful.

I share your concern. By the time the query is in baseplate, it's already a string or SimpleStatement type. With that in mind, there's no structured way I can see to access things like key or column family without parsing the CQL statement.

Reply:
> By the time the query is in baseplate, it's already a string or SimpleStatement type.

That's the external API. By the time it's written in an application it's like that. This is pretty similar to how we don't tag tables etc. in SQL because it'd just require writing a full query parser for it.

@bsimpson63 do you have any thoughts on what benefit extracting and tagging this stuff has? I understand r2 has it, but that's the thrift API and I'm just not sure how worthwhile this is.

Either way, if we want to move forward with this I'd like to see some performance testing gauge the impact, pick out any low-hanging fruit of improvements, and potentially inform the choice of ditching the idea, sampling, or going full-bore.

Reply:
I talked to @ckwang8128. The aim here was to allow developers to search for traces based on column family and statement type. We're going to see if we can't provide this another way. Possibly by making improving the query functionality of the trace store by moving to ES.